presto-spi/src/main/java/com/facebook/presto/spi/SplitWeight.java

```diff
@@ -25,6 +26,7 @@
 public final class SplitWeight
 {
     private static final long UNIT_VALUE = 100;
```

Comment:
I really don't think using long is a good idea. 

First of all, by using long and UNIT_VALUE 100, it loses precision and finer granularity since it can only represent 100 granules. 

Secondly, this would regress performance instead of improving it as you originally thought. Since you used long in SPI and double in connector, you had to convert the value between the two:
```
public static SplitWeight fromProportion(double weight)
{
    return fromRawValue((long) Math.ceil(weight * UNIT_VALUE));
}
```
The use of Math.ceil() is in fact much more expensive than ADD instructions on double precision floating numbers. An ADD on double takes 4 cycles on Intel CPU (Nehalem) while ceil() or floor() takes ten's or hundred's of cycles: https://stackoverflow.com/questions/23203710/floating-point-operations-per-cycle-intel . You also had to do a multiplication `weight * UNIT_VALUE` instead of a simple assignment, and multiplication is more expensive than ADD  on double or integer types on most CPUs. All these don't need to happen if you just use double in both systems, and you can simply pass the value from connector to scheduler, and the only arithmetic operation you needed to do is mostly ADD. It should be much faster than this conversion. Furthermore, the double arithmetic is the least thing you need to worry about in this scenario. Listing directory is many orders of magnitude higher than adding some, even millions of doubles. In the past I never saw the split assignment related operations would be CPU bottlenecks. You are welcome to write a JMH benchmark to test this. 

Thirdly, it added extra logic and complexity to the code for maintaining two weight systems. Code becomes more complex, and developers get confused.

Based on the above, I think having two weight types is a bad idea. I didn't know why Trino side suggested this. Is there something I didn't know? If not any particular reasons, I would strongly suggest unify the two weighting systems.

Reply:
Let me try to walk through the thinking about how we ended up with this version of `SplitWeight` representations in the original PR and then through the changes made in the course of the Trino PR revisions.

First, I want to highlight the different places and characteristics of where these weight values are used:

1. The SPI: Each split must be able to assign a weight value, and connectors not supporting split weighting must have a default "standard" weight. Ideally, the "standard" weight and general weight calculation system should be intuitive and avoid creating compatibility headaches in the future.
2. The NodeScheduler and supporting classes like `NodeTaskMap`, `NodeAssignmentStats`, and split placement logic. This logic does a few things that start to create some real constraints on possible solutions:
- The coordinators view of a current task must track the combined weight of splits that are queued locally, queued remotely, and running splits separately performs arithmetic on those separate weight sums to combine the queued locally weight with the weights from the last observed remote task status update.
- The task level accounting then contributes its computed view of the current task via and deltas from previous values into the `NodeTaskMap`, which tracks the current total weight running on a node through these incrementally computed deltas for all tasks on a given node, and does so over the course of that nodes entire lifetime.
- Failing to accurately account weights is a risk, and introduces the requirement that split weight arithmetic must not lose accuracy in the course of performic arithmetic over arbitrarily ordered additions and subtractions of combined sums and deltas between them, otherwise a node's state would drift away from the "correct" value and cause bad scheduling decisions in either direction (ie: node is "full" when it's actually not running anything, or "empty" when it's actually fully loaded).
- Split scheduling is latency sensitive, especially when splits are small and the task / node state tracking updates all occur within critical sections or compareAndSet operations. Overhead of calculation should be minimized where possible. For what it's worth, I think `double` would be fine here, but `BigDecimal` would be a performance risk.
3. Task and pipeline status updates inside of split assignments must also accumulate and perform arithmetic operations to determine the worker's view of currently queued / running / completed weights which is similar to the above coordinator side problem. It's also similar in that these accounting points happen in critical paths that are expected to be somewhat contended. The final bit of complexity that the task status updates provide however, is that however the state is computed and represented- it must be able to represent the current computed state value back to the coordinator in JSON format. This causes more problems because floating point values must lose at least _some_ precision when serialized to strings (ie: 0.33333...) and might need a large number of digits to lost only an _acceptable_ amount of precision.


With the above constraints, I chose to use a single `long` field  in the first iteration of this PR with a "standard" value of 100. The choice solved problems associated with items 2 and 3 above, because arbitrarily ordered arithmetic operations over single and aggregated weights is not a problem with integer arithmetic, calculations are cheap to perform, and values serialize correctly and succinctly in JSON. `double` would probably perform sufficiently well, but risked precision loss beyond just that introduced by serialization. `BigDecimal` would probably have been able to reach a reasonable amount of precision such that even with JSON serialization the loss was negligible, but would have been a performance risk. This unfortunately, forced me to compromised on item 1 initially.

In the Trino PR the initial pushback on this attempt was similar:
- Computing weights "relative to 100" is awkward and unintuitive- computing them relative to 1.0 is more natural
- Exposing a magic hard-coded constant of 100 as a "standard" value would be cemented into the SPI contract after that point and hard to change
- What if 100 "granules" isn't enough and we want to be able to express weights at higher resolution (eg: 1,000 or 1M)

The resolution was what you see updated in the latest version of this PR:
- The internal representation remains a `long` to avoid loss of accuracy in arithmetic and serialization
- The preferred construction of weight values for connectors is via `SplitWeight.fromProportion(double)` which is more intuitive, but allows changes in the granularity in the future if necessary


That kind of leaves the last remaining item here: is 100 _enough_ granularity? In my experience with this setting, yes. Remember, the weighting mechanism is designed to make the coordinator split assignment rate adaptive to avoid bottlenecking on split scheduling latency which leavees workers underutilized. Once workers are given sufficiently many splits to stay busy, the goal is accomplished and expressing and no further granularity is especially important. This is incidentally part of the reason which is why the default "hive.minimum-assigned-split-weight" is 0.05 (1/20 of a standard split) or 20x more splits allowed, and not 100x more splits. Different environments might have different needs, so if in the future we decide that 1,000 or 1M should become the definition of "standard", that's still an option with the solution in this current PR.